File: 1650_Lowest Common Ancestor of a Binary Tree III.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node parent;
};
*/

// Method 1: hashset
class Solution {
    public Node lowestCommonAncestor(Node p, Node q) {
        if (p == q.parent)
            return p;
        if (q == p.parent)
            return q;
        if (p.parent == q.parent)
            return p.parent;

        HashSet<Node> set = new HashSet<>();

        while (p.parent != null) {
            set.add(p);
            p = p.parent;
        }

        while (q.parent != null && !set.contains(q)) {
            q = q.parent;
        }
        return q;
    }
}

// Method 2: align depth
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node parent;
};
*/

class Solution {
    public Node lowestCommonAncestor(Node p, Node q) {
        int pDepth = getDepth(p), qDepth = getDepth(q);

        Node deep = (pDepth >= qDepth) ? p : q;
        Node shallow = (pDepth < qDepth) ? p : q;
        int diff = Math.abs(pDepth - qDepth);

        while (diff > 0) {
            deep = deep.parent;
            diff--;
        }

        while (deep != shallow) {
            deep = deep.parent;
            shallow = shallow.parent;
        }
        return deep;
    }

    private int getDepth(Node n) {
        int depth = 0;
        while (n != null) {
            depth++;
            n = n.parent;
        }
        return depth;
    }
}

File: 1650_Lowest Common Ancestor of a Binary Tree III.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
        self.parent = None
"""

class Solution:
    def lowestCommonAncestor(self, p: 'Node', q: 'Node') -> 'Node':
        """
        Let a = p -> lca, b = q -> lca, c = lca -> head.
        ptrP = a + c + b, ptrQ = b + c + a (switch to another node)
        Same as LC160
        """
        ptrP, ptrQ = p, q

        while ptrP != ptrQ:
            ptrP = ptrP.parent if ptrP else q
            ptrQ = ptrQ.parent if ptrQ else p
        
        return ptrP